NeoML/include/NeoML/Dnn/Layers/MobileNetV2BlockLayer.h

```diff
@@ -0,0 +1,108 @@
+	CPtr<CDnnBlob> GetExpandFilter() const { return getParamBlob( P_ExpandFilter ); }
+	void SetExpandFilter( const CPtr<CDnnBlob>& blob ) { setParamBlob( P_ExpandFilter, blob ); }
+	CPtr<CDnnBlob> GetExpandFreeTerm() const { return getParamBlob( P_ExpandFreeTerm ); }
+	void SetExpandFreeTerm( const CPtr<CDnnBlob>& blob ) { setParamBlob( P_ExpandFreeTerm, blob ); }
+	float GetExpandReLUThreshold() const { return expandReLUThreshold.GetValue(); }
+	void SetExpandReLUThreshold( float newValue ) { expandReLUThreshold.SetValue( newValue ); }
+
+	// Channelwise convolution and ReLU parameters
+	int GetStride() const { return stride; }
+	void SetStride( int newValue ) { NeoAssert( newValue == 1 || newValue == 2 ); stride = newValue; }
+	CPtr<CDnnBlob> GetChannelwiseFilter() const { return getParamBlob( P_ChannelwiseFilter ); }
+	void SetChannelwiseFilter( const CPtr<CDnnBlob>& blob ) { setParamBlob( P_ChannelwiseFilter, blob ); }
+	CPtr<CDnnBlob> GetChannelwiseFreeTerm() const { return getParamBlob( P_ChannelwiseFreeTerm ); }
+	void SetChannelwiseFreeTerm( const CPtr<CDnnBlob>& blob ) { setParamBlob( P_ChannelwiseFreeTerm, blob ); }
+	float GetChannelwiseReLUThreshold() const { return channelwiseReLUThreshold.GetValue(); }
+	void SetChannelwiseReLUThreshold( float newValue ) { channelwiseReLUThreshold.SetValue( newValue ); }
+
+	// Down convolution parameters
+	CPtr<CDnnBlob> GetDownFilter() const { return getParamBlob( P_DownFilter ); }
+	void SetDownFilter( const CPtr<CDnnBlob>& blob ) { setParamBlob( P_DownFilter, blob ); }
+	CPtr<CDnnBlob> GetDownFreeTerm() const { return getParamBlob( P_DownFreeTerm ); }
+	void SetDownFreeTerm( const CPtr<CDnnBlob>& blob ) { setParamBlob( P_DownFreeTerm, blob ); }
+
+	// Residual connection
+	bool HasResidual() const { return residual; }
+	void SetResidual( bool newValue ) { residual = newValue; }
+
+	// Serialization
+	void Serialize( CArchive& archive ) override;
+
+protected:
+	// CBaseLayer methods
+	void Reshape() override;
+	void RunOnce() override;
+	void BackwardOnce() override { NeoAssert( false ); }
+
+private:
+	// paramBlobs indices
+	enum TParam {
+		P_ExpandFilter,
+		P_ExpandFreeTerm,
+		P_ChannelwiseFilter,
+		P_ChannelwiseFreeTerm,
+		P_DownFilter,
+		P_DownFreeTerm,
+
+		P_Count
+	};
+
+	bool residual; // Does block have residual connection?
+	int stride; // stride of channnelwise convolution
+	CFloatHandleVar expandReLUThreshold; // threshold of expand convolution ReLU
+	CFloatHandleVar channelwiseReLUThreshold; // threshold of channelwise convolution ReLU
+	CChannelwiseConvolutionDesc* convDesc; // descriptor of channelwise convolution
+
+	CPtr<CDnnBlob> getParamBlob( TParam param ) const;
+	void setParamBlob( TParam param, const CPtr<CDnnBlob>& blob );
+};
+
+// Optimizes MobileNetV2-like net by replacing blocks of layers with CMobileNetV2BlockLayer
+// Improves inference performance but make net untraninable
+int NEOML_API OptimizeMobileNetV2( CDnn& dnn );
```

Comment:
What does it return? Maybe bool?

Reply:
I forgot to mention. It returns the number of blocks replaced.

I definitely don't like the idea of `bool` because it's not informative. See `CDnnOptimizationReport` struct.